### PR TITLE
6763 Fix FED datetime format for API submission

### PR DESCRIPTION
**Pull request by argush3**
Issue #: /bcgov/entity#6763

Description of changes:

- fixed FED datetime format for API submission
- added unit test

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Comment by severinbeauvais** [comment]
Please update the title to something shorter but informative.

Please update the description block according to the template. See other PRs as example (eg, https://github.com/bcgov/business-edit-ui/pull/138).

You may not have permissions to do so, but this PR should be assigned to YOU and appropriate reviewers should be assigned. (I will do it this time.)

File: src/mixins/date-mixin.ts [hunk 1]
```diff
@@ -187,4 +208,15 @@ export default class DateMixin extends Mixins(CommonMixin) {
     const dateLocalMs = new Date(date).setHours(0, 0, 0, 0)
     return Math.round((dateLocalMs - todayLocalMs) / this.MS_IN_A_DAY)
   }
+
+  /**
+   * Padding function to ensure month, date, hour, minutes and seconds front padded with zero if input val < 10
+   * @returns '01' for 1
+   * @returns 10 for 10
+   */
+  pad (number : Number) : String {
```
Comment (severinbeauvais) [inline src/mixins/date-mixin.ts:217]: The formal parameter type should be `number` (which is a type) not `Number` (which is a class).
The return type should be fixed similarly.

Finally, you don't need this as there is already a JS equivalent: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart

File: src/mixins/date-mixin.ts [hunk 2]
```diff
@@ -171,6 +171,27 @@ export default class DateMixin extends Mixins(CommonMixin) {
...
+    // convert date object to ISO format that API accepts
+    // eg, 2021-03-25T10:02:00.000000+00:00
+    const dateStr = date.getUTCFullYear() +
+      '-' + this.pad(date.getUTCMonth() + 1) +
+      '-' + this.pad(date.getUTCDate()) +
+      'T' + this.pad(date.getUTCHours()) +
+      ':' + this.pad(date.getUTCMinutes()) +
+      ':' + this.pad(date.getUTCSeconds()) +
+      '.' + (date.getUTCMilliseconds() / 1000).toFixed(6).slice(2, 8) +
+      '+00:00'
```
Comment (severinbeauvais) [inline src/mixins/date-mixin.ts:190]: In my opinion, this is not the correct solution.

Admittedly, both the output of `date.toISOString()` (eg, 2021-01-02T07:59:59.000Z) and the Python format (eg, 2021-01-02T07:59:59.000Z) are ISO 8601, but apparently aren't compatible.

Given the examples above, it should be trivial to convert from one to the other. I would simply knock off the "Z" and append the offset string. See apiToDateAndTimeString() above and reverse the process.

Comment (argush3) [reply src/mixins/date-mixin.ts:190]: so if i re-work this function(`dateToApiIsoDateTimeString`) to use `date.toISOString()` as a starting point, I would still be missing part of the milliseconds that the API expects.  What do we want to do about this?  Should we replace the Z with 000+00:00 ?  Or should I just grab the milliseconds with what I was originally using?  i.e. `(date.getUTCMilliseconds() / 1000).toFixed(6).slice(2, 8)`

Comment (severinbeauvais) [reply src/mixins/date-mixin.ts:190]: As it turns out, `toISOString()` gives you the milliseconds, eg
```
> new Date().toISOString()
> "2021-03-22T20:20:56.052Z"
```

How inflexible is the Python ISO date parser? Do you have specify it all exactly? If so then, yes, you'll have to right-pad the milliseconds (and add the tz offset).

This is what I would do (but feel free to do something nicer!):
```
const dateStr = date.toISOString().slice(1,23) + '000+00:00'
```

Ref: https://www.w3schools.com/jsref/jsref_slice_array.asp

Comment (severinbeauvais) [reply src/mixins/date-mixin.ts:190]: ^^ and add a code comment describing what you're doing ;)

Comment (argush3) [reply src/mixins/date-mixin.ts:190]: For some reason I thought I needed the additional milliseconds.  I just tested it and the api did accept a format of `2021-03-26T13:04:00.000Z`.  I might have been just thinking we might need the millisecond precision part which is highly unlikely if I think about it.

So I'll just do as per your suggestion:

```
const dateStr = date.toISOString().slice(1,23) + '+00:00'
```

**Review by severinbeauvais — COMMENTED** [review]
The code looks fine.

One last thing (sorry, I should have mentioned it sooner)... Please update the app's version by 0.0.1 in package.json (and then do `npm i`), and check in the new changes. This is to easily tell what version is deployed.

**Review by severinbeauvais — APPROVED** [review]

**Pull request merged**